{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e7537753",
   "metadata": {},
   "source": [
    "# This is a notebook to generate Deep mutation scanning library for scoring sequence using ESM_if based on protein backbone\n",
    "\n",
    "To order to compare computational and expeirmental results,\n",
    "we need to match sequences\n",
    "2. We need use mask in cooridnates for sites where pdb cooridnates are not avaialale\n",
    "\n",
    "The following is an application to SHP2 protein."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a387c0fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/yunyao/miniconda3/envs/struct-evo/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/home/yunyao/miniconda3/envs/struct-evo/lib/python3.9/site-packages/esm/pretrained.py:215: UserWarning: Regression weights not found, predicting contacts will not produce correct results.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import esm\n",
    "#load the model\n",
    "model, alphabet = esm.pretrained.esm_if1_gvp4_t16_142M_UR50()\n",
    "model = model.eval()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "841d0081",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Native sequence:\n",
      "SRRWFHPNITGVEAENLLLTRGVDGSFLARPSKSNPGDFTLSVRRNGAVTHIKIQNTGDYYDLYGGEKFATLAELVQYYMEHHGQLKEKNGDVIELKYPLNCADPTSERWFHGHLSGKEAEKLLTEKGKHGSFLVRESQSHPGDFVLSVRTGDDKGESNDGKSKVTHVMIRCQELKYDVGGGERFDSLTDLVEHYKKNPMVETLGTVLQLKQPLNTTRINAAEIESRVRELSKGFWEEFETLQQQECKLLYSRKEGQRQENKNKNRYKNILPFDHTRVVLHDSDYINANIIMPEKKSYIATQGCLQNTVNDFWRMVFQENSRVIVMTTKEVERGKSKCVKYWPDEYALKEYGVMRVRNVKESAAHDYTLRELKLSKVGQGNTERTVWQYHFRTWPDHGVPSDPGGVLDFLEEVHHKQESIMDAGPVVVHCSAGIGRTGTFIVIDILIDIIREKGVDCDIDVPKTIQMVRSQRSGMVQTEAQYRFIYMAVQHYIETLQRR\n",
      "Coordinates shape: (499, 3, 3)\n"
     ]
    }
   ],
   "source": [
    "# information on structure to be designed\n",
    "fpath = './4DGP.pdb' # .pdb format is also acceptable\n",
    "chain_id = 'A'\n",
    "structure = esm.inverse_folding.util.load_structure(fpath, chain_id)\n",
    "coords, native_seq = esm.inverse_folding.util.extract_coords_from_structure(structure)\n",
    "print('Native sequence:')\n",
    "print(native_seq)\n",
    "print('Coordinates shape:', coords.shape)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d9d420f",
   "metadata": {},
   "source": [
    "##  Warning!! The above native sequence is not correct.\n",
    "\n",
    "The model ignores residue of which coordiantes are not available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9587f59d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "dir_path = '/home/yunyao/structural-evolution/bin'\n",
    "\n",
    "# Add the directory containing the modules to sys.path\n",
    "if dir_path not in sys.path:\n",
    "    sys.path.append(dir_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7874a239",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we can test the fitnesss of various sequences generated by dms\n",
    "\n",
    "import argparse\n",
    "from dms_utils import deep_mutational_scan\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "\n",
    "import esm\n",
    "from util import load_structure, extract_coords_from_structure\n",
    "import biotite.structure\n",
    "from collections import defaultdict\n",
    "\n",
    "\n",
    "def get_native_seq(pdbfile, chain):\n",
    "    structure = load_structure(pdbfile, chain)\n",
    "    _ , native_seq = extract_coords_from_structure(structure)\n",
    "    return native_seq\n",
    "\n",
    "\n",
    "def read_fasta(file_path):\n",
    "    '''Reads a FASTA file and returns a dictionary with sequence IDs as keys and sequences as values.'''\n",
    "    sequences = {}\n",
    "    with open(file_path, 'r') as f:\n",
    "        seq_id = None\n",
    "        seq = ''\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if seq_id is not None:\n",
    "                    sequences[seq_id] = seq\n",
    "                seq_id = line[1:5]  # Remove '>'\n",
    "                seq = ''\n",
    "            else:\n",
    "                seq += line\n",
    "        if seq_id is not None:\n",
    "            sequences[seq_id] = seq  # Add the last sequence\n",
    "    return seq\n",
    "    \n",
    "def write_dms_lib(args):\n",
    "    '''Writes a deep mutational scanning library, including the native/wildtype (wt) of the \n",
    "    indicated target chain in the structure to an output Fasta file'''\n",
    "\n",
    "    sequence = ''.join(read_fasta(args.seqfile).split())\n",
    "    Path(args.dmspath).parent.mkdir(parents=True, exist_ok=True)\n",
    "    with open(args.dmspath, 'w') as f:\n",
    "        f.write('>wt\\n')\n",
    "        f.write(sequence+'\\n')\n",
    "        for pos, wt, mt in deep_mutational_scan(sequence):\n",
    "            assert(sequence[pos] == wt)\n",
    "            mut_seq = sequence[:pos] + mt + sequence[(pos + 1):]\n",
    "\n",
    "            f.write('>' + str(wt) + str(pos+1) + str(mt) + '\\n')\n",
    "            f.write(mut_seq + '\\n')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ce5b11b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GMTSRRWFHPNITGVEAENLLLTRGVDGSFLARPSKSNPGDFTLSVRRNGAVTHIKIQNTGDYYDLYGGEKFATLAKLVQYYMEHHGQLKEKNGDVIELKYPLNCADPTSERWFHGHLSGKEAEKLLTEKGKHGSFLVRESQSHPGDFVLSVRTGDDKGESNDGKSKVTHVMIRCQELKYDVGGGERFDSLTDLVEHYKKNPMVETLGTVLQLKQPLNTTRINAAEIESRVRELSKLAETTDKVKQGFWEEFETLQQQECKLLYSRKEGQRQENKNKNRYKNILPFDHTRVVLHDGDPNEPVSDYINANIIMPEFETKCNNSKPKKSYIATQGCLQNTVNDFWRMVFQENSRVIVMTTKEVERGKSKCVKYWPDEYALKEYGVMRVRNVKESAAHDYTLRELKLSKVGQGNTERTVWQYHFRTWPDHGVPSDPGGVLDFLEEVHHKQESIMDAGPVVVHCSAGIGRTGTFIVIDILIDIIREKGVDCDIDVPKTIQMVRSQRSGMVQTEAQYRFIYMAVQHYIETL'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test the read_fasta function\n",
    "test_seqs= read_fasta('/home/yunyao/SH2/6CRF.fasta')\n",
    "test_seq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ac9a2897",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdbfile_path= './4DGP.pdb'\n",
    "chain_id='A'\n",
    "outpath='./dms_lib2'\n",
    "dmspath = f'{outpath}/4DGP_dms_lib2.fasta'\n",
    "seqfile_path = '/home/yunyao/SH2/4DGP.fasta'\n",
    "write_dms_lib(\n",
    "    argparse.Namespace(\n",
    "        seqfile=seqfile_path,\n",
    "        pdbfile=pdbfile_path,\n",
    "        chain=chain_id,\n",
    "        dmspath=dmspath,\n",
    "        outpath=outpath\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f23ca897",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we will remove the first residue so that it matches experimental sequence\n",
    "# for simplicity, we remove the G directly in fasta file\n",
    "pdbfile_path= './6CRF.pdb'\n",
    "chain_id='A'\n",
    "outpath='./dms_lib2'\n",
    "dmspath = f'{outpath}/6CRF_dms_lib2.fasta'\n",
    "seqfile_path = '/home/yunyao/SH2/6CRF.fasta'\n",
    "write_dms_lib(\n",
    "    argparse.Namespace(\n",
    "        seqfile=seqfile_path,\n",
    "        pdbfile=pdbfile_path,\n",
    "        chain=chain_id,\n",
    "        dmspath=dmspath,\n",
    "        outpath=outpath\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80e68cc0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligned Sequence 1: MTSRRWFHPNITGVEAENLLLTRGVDGSFLARPSKSNPGDFTLSVRRNGAVTHIKIQNTGDYYDLYGGEKFATLAELVQYYMEHHGQLKEKNGDVIELKYPLNCADPTSERWFHGHLSGKEAEKLLTEKGKHGSFLVRESQSHPGDFVLSVRTGDDKGESNDGKSKVTHVMIRCQELKYDVGGGERFDSLTDLVEHYKKNPMVETLGTVLQLKQPLNTTRINAAEIESRVRELSKLAETTDKVKQGFWEEFETLQQQECKLLYSRKEGQRQENKNKNRYKNILPFDHTRVVLHDGDPNEPVSDYINANIIMPEFETKCNNSKPKKSYIATQGCLQNTVNDFWRMVFQENSRVIVMTTKEVERGKSKCVKYWPDEYALKEYGVMRVRNVKESAAHDYTLRELKLSKVGQGNTERTVWQYHFRTWPDHGVPSDPGGVLDFLEEVHHKQESIMDAGPVVVHCSAGIGRTGTFIVIDILIDIIREKGVDCDIDVPKTIQMVRSQRSGMVQTEAQYRFIYMAVQHYIETLQRRLEHHHHHH\n",
      "Aligned Sequence 2: --SRRWFHPNITGVEAENLLLTRGVDGSFLARPSKSNPGDFTLSVRRNGAVTHIKIQNTGDYYDLYGGEKFATLAELVQYYMEHHGQLKEKNGDVIELKYPLNCADPTSERWFHGHLSGKEAEKLLTEKGKHGSFLVRESQSHPGDFVLSVRTGDDKGESNDGKSKVTHVMIRCQELKYDVGGGERFDSLTDLVEHYKKNPMVETLGTVLQLKQPLNTTRINAAEIESRVRELS---------K-GFWEEFETLQQQECKLLYSRKEGQRQENKNKNRYKNILPFDHTRVVLH--D-----SDYINANIIMP--E--------KKSYIATQGCLQNTVNDFWRMVFQENSRVIVMTTKEVERGKSKCVKYWPDEYALKEYGVMRVRNVKESAAHDYTLRELKLSKVGQGNTERTVWQYHFRTWPDHGVPSDPGGVLDFLEEVHHKQESIMDAGPVVVHCSAGIGRTGTFIVIDILIDIIREKGVDCDIDVPKTIQMVRSQRSGMVQTEAQYRFIYMAVQHYIETLQRR--------\n",
      "Mask: [0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "operands could not be broadcast together with shapes (536,1) (499,3,3) () ",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[29], line 41\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMask:\u001b[39m\u001b[38;5;124m\"\u001b[39m, mask)\n\u001b[1;32m     38\u001b[0m \u001b[38;5;66;03m# use the mask to generate new cooridnates\u001b[39;00m\n\u001b[1;32m     39\u001b[0m \u001b[38;5;66;03m# when maks is 1, we keep the coordinates\u001b[39;00m\n\u001b[1;32m     40\u001b[0m \u001b[38;5;66;03m# when mask is 0, we add np.inf to the coordinates\u001b[39;00m\n\u001b[0;32m---> 41\u001b[0m maksed_coords_4DGP\u001b[38;5;241m=\u001b[39m \u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwhere\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43marray\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmask\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcoords\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43minf\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     42\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMasked coordinates shape:\u001b[39m\u001b[38;5;124m'\u001b[39m, maksed_coords_4DGP\u001b[38;5;241m.\u001b[39mshape)\n",
      "File \u001b[0;32m<__array_function__ internals>:180\u001b[0m, in \u001b[0;36mwhere\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: operands could not be broadcast together with shapes (536,1) (499,3,3) () "
     ]
    }
   ],
   "source": [
    "# now we need to generate coordinates mask\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "# Function to align sequences and generate a mask\n",
    "def align_sequences(seq1, seq2):\n",
    "    # Perform global alignment using the Needleman-Wunsch algorithm\n",
    "    alignments = pairwise2.align.globalxx(seq1, seq2)\n",
    "\n",
    "    # Take the best alignment (highest score)\n",
    "    best_alignment = alignments[0]\n",
    "    aligned_seq1, aligned_seq2, score, start, end = best_alignment\n",
    "\n",
    "    # Generate the mask\n",
    "    mask = []\n",
    "    for a, b in zip(aligned_seq1, aligned_seq2):\n",
    "        if a == b and a != \"-\":\n",
    "            mask.append(1)  # Match\n",
    "        else:\n",
    "            mask.append(0)  # Mismatch or gap\n",
    "\n",
    "    return aligned_seq1, aligned_seq2, mask\n",
    "\n",
    "# Example usage\n",
    "full_4DGP = ''.join(read_fasta('/home/yunyao/SH2/4DGP.fasta').split())\n",
    "full_6CRF = ''.join(read_fasta('/home/yunyao/SH2/6CRF.fasta').split())\n",
    "Coor_seq_4DGP= get_native_seq('./4DGP.pdb', 'A')\n",
    "Coor_seq_6CRF= get_native_seq('./6CRF.pdb', 'A')\n",
    "\n",
    "\n",
    "aligned_seq1, aligned_seq2, mask = align_sequences(full_4DGP, Coor_seq_4DGP)\n",
    "\n",
    "# Output the results\n",
    "print(\"Aligned Sequence 1:\", aligned_seq1)\n",
    "print(\"Aligned Sequence 2:\", aligned_seq2)\n",
    "print(\"Mask:\", mask)\n",
    "\n",
    "# use the mask to generate new cooridnates\n",
    "# when maks is 1, we keep the coordinates\n",
    "# when mask is 0, we add np.inf to the coordinates\n",
    "# maked_coords_4DGP shous have shape as (len(mask), 3, 3)\n",
    "# when maask is 1, fill it with the coordinates\n",
    "# when mask is 0, fill it with np.inf\n",
    "import numpy as np  \n",
    "# Assuming coords is a numpy array of shape (N, 3, 3) where N is the number of residues\n",
    "fpath = './4DGP.pdb' # .pdb format is also acceptable\n",
    "chain_id = 'A'\n",
    "structure = esm.inverse_folding.util.load_structure(fpath, chain_id)\n",
    "coords, native_seq = esm.inverse_folding.util.extract_coords_from_structure(structure)\n",
    "\n",
    "#fill coords into masked_coords_4DGP\n",
    "maksed_coords_4DGP=\n",
    "\n",
    "np.where(np.array(mask)[:, None], coords, np.inf)\n",
    "print('Masked coordinates shape:', maksed_coords_4DGP.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "91ebc061",
   "metadata": {},
   "outputs": [],
   "source": [
    "  # now we can scan through the library and score each sequence\n",
    "dms_lib = defaultdict(list)\n",
    "with open(dmspath, 'r') as f:\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            seq_id = line.strip()[1:]  # remove '>'\n",
    "        else:\n",
    "            sequence = line.strip()\n",
    "            dms_lib[seq_id].append(sequence)\n",
    "# score each sequence in the library\n",
    "scores = {}\n",
    "for seq_id, sequences in dms_lib.items():\n",
    "    for sequence in sequences:\n",
    "        ll_fullseq, ll_withcoord = esm.inverse_folding.util.score_sequence(model, alphabet, coords, sequence)\n",
    "        scores[seq_id] = (ll_fullseq, ll_withcoord)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d92a2367",
   "metadata": {},
   "outputs": [],
   "source": [
    "python ~/structural-evolution/bin/score_log_likelihoods.py ./4DGP.pdb  \\\n",
    "    ./dms_lib/4DGP_dms_lib.fasta --chain A \\\n",
    "    --outpath output/4DGP_mutated_seqs_scores.csv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "struct-evo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
